### ISSUE TITLE
## RFC: Figma design system → React/TypeScript adaptor pack

### ISSUE THREAD

**Issue by falese** [falese/coder#22]
## RFC: Figma design system → React/TypeScript adaptor pack

**Type:** Feature / RFC
**Phase:** Backlog (unscheduled — depends on Phase 5 completion)
**Labels:** `rfc`, `adaptor-pack`, `backlog`

-----

### Summary

Add a `figma-to-react` adaptor pack that ingests a Figma design system toolkit via the Figma REST API and generates production-quality React/TypeScript components mapped to your existing component library and design tokens.

This is distinct from the React/TS adaptor (#11) — that adaptor learns React patterns from open source code. This adaptor learns to translate a specific design system’s token and component vocabulary into its corresponding codebase implementation.

-----

### Motivation

When a project has a complete Figma design system toolkit (tokens, grids, component library with variants and states), the gap between design and code is primarily a mapping problem — not a reasoning problem. Design tokens resolve named values (`color.primary.600`) to named implementations (`bg-primary-600`). Component references in Figma frames map to existing React components by name convention. Layout properties (auto-layout, constraints, breakpoints) map deterministically to Flexbox and responsive classes.

A LoRA adaptor trained on prompt/completion pairs derived from this mapping can close that gap with high consistency, without requiring a vision model or cloud inference.

-----

### Feasibility notes

- **Does not require a vision/multimodal model.** The Figma REST API returns the full node tree as JSON with all style references already resolved to token names. The model never sees pixel values or hex colors — only named tokens.
- **Fits the existing 7B + r=8 LoRA stack.** The mapping problem is tight and well-defined — exactly the kind of domain a small LoRA handles well.
- **Training data is unusually high quality.** Completions are your actual production components, not synthetic output. Ground truth is already written.
- **Layout inference is deterministic, not learned.** Auto-layout → Flexbox mapping is handled by the ingestion pipeline as pre-processing, not trained into the LoRA.

-----

### Prompt / completion format

Each training pair represents one component boundary. The prompt is built from the Figma node tree with all style references resolved to token names. The completion is the existing React/TypeScript implementation.

**Prompt structure:**

```
COMPONENT: Card/Product
VARIANT: elevated
LAYOUT: column gap=spacing.4 padding=spacing.6
TOKENS:
  background=color.surface.primary
  border=color.border.subtle
  radius=radius.lg
CHILDREN:
  - Image/Thumbnail [aspectRatio=16:9 width=fill]
  - Text/Heading.sm [token=color.text.primary truncate=true]
  - Text/Body.md [token=color.text.secondary lines=2]
  - Button/Primary [label=dynamic size=sm]
STATE: default hover disabled
RESPONSIVE: stack at breakpoint.sm
```

**Completion:** the actual TypeScript/React component from your codebase — imports, props interface, JSX.

**Format rules:**

- Raw Figma values (hex, px) are never in the prompt — resolved to token names by the pipeline
- `COMPONENT:` uses the Figma path as the lookup key (e.g. `Card/Product`)
- `CHILDREN:` lists Figma component references, not raw frame descriptions
- `STATE:` is derived from Figma variant properties
- `RESPONSIVE:` is included only when the design system encodes breakpoint annotations
- Complex compositions are decomposed into one pair per component boundary — never one prompt per page

-----

### Ingestion pipeline (new deliverable)

The pipeline is the core new work. It needs to:

1. Authenticate with the Figma REST API (`FIGMA_ACCESS_TOKEN`)
1. Walk the published component library and resolve all style references to token names
1. Match each Figma component to its codebase implementation via name convention or an explicit manifest
1. Extract the implementation file as the completion
1. Build the structured prompt from the resolved node tree
1. Handle decomposition of nested compositions into individual component pairs
1. Write `train.jsonl` + `eval.jsonl` (85/15 split)

The pipeline is invoked via `coder data ingest --figma <file-key> --map <component-manifest>` and produces a JSONL dataset compatible with the standard `coder adaptor train` workflow.

-----

### Component manifest

A mapping file that ties Figma component paths to codebase import paths, for cases where naming convention alone isn’t sufficient:

```json
{
  "Card/Product": "@/components/cards/ProductCard",
  "Button/Primary": "@/components/Button",
  "Image/Thumbnail": "@/components/Thumbnail"
}
```

-----

### Adaptor pack structure

Follows the standard pack schema with one addition — a `figma/` directory for design system metadata:

```
figma-to-react/
├── weights/
│   └── adaptor.safetensors
├── data/
│   ├── train.jsonl
│   └── eval.jsonl
├── evals/
│   └── eval_suite.ts
├── prompts/
│   └── system.md
├── figma/
│   ├── component-manifest.json   # Figma path → codebase import mapping
│   └── token-map.json            # Token name → CSS class / variable mapping
└── manifest.json
```

-----

### Eval suite

Beyond the standard tsc / ESLint / test-pass-rate dimensions, the figma adaptor eval should additionally check:

- **Token coverage** — all tokens referenced in the prompt appear in the generated output
- **Component reuse** — generated code imports existing components, does not reimplement them
- **No raw values** — no hex colors, magic numbers, or hardcoded pixel values in output

-----

### Dependencies

- #11 React/TS adaptor pack (shares component library conventions and eval infrastructure)
- `coder data ingest` pipeline (#7) extended with `--figma` flag
- Figma REST API access token (user-supplied, stored in `~/.coder/config.toml` as `figma_token`)

-----

### Out of scope for this issue

- Screenshot / image → code (requires vision model, not feasible on 18GB)
- Multi-file page-level generation (one component at a time only)
- Figma Dev Mode integration (API-based only)
- Cross-project design system resolution

-----

### Open questions

- [ ] Should the component manifest be hand-authored or auto-generated from the Figma component library + codebase name matching?
- [ ] How are Figma component variants (e.g. `Button/Primary/Hover`) handled — separate pairs or a single pair with STATE field?
- [ ] Token map format: dot-notation → Tailwind class, CSS variable, or both?
